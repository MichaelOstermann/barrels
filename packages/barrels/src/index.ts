// Generated by @monstermann/barrels
/* eslint-disable */
/* prettier-ignore */
// dprint-ignore-file
// biome-ignore lint: disable

export * from "./Barrel";
export * from "./defineConfig";
export * from "./Source";
export * from "./SourceModule";
export * from "./types";
