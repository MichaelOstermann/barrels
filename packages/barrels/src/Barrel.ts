import type { BarrelsOutput } from "./types"
import { mkdir, writeFile } from "node:fs/promises"
import Path from "node:path"
import { createFilter } from "unplugin-utils"
import { getBarrelsOutput } from "./internals/context"
import { Source } from "./Source"

const barrelsCache = new Map<string, string>()

export const Barrel = {
    ignoresHeader: `// Generated by @monstermann/barrels
/* eslint-disable */
/* prettier-ignore */
// dprint-ignore-file
// biome-ignore lint: disable`,
    constExport({ name, type, value }: {
        name: string
        type?: boolean
        value: string
    }): string {
        if (!name || !value) return ""
        return [
            "export",
            type ? "type" : "const",
            name,
            "=",
            `${value};`,
        ]
            .join(" ")
    },
    defaultExport({ alias, path, type }: {
        alias?: string
        path: string
        type?: boolean
    }): string {
        if (!path) return ""
        return [
            "export",
            type ? "type" : "",
            alias ? `{ default as ${alias} }` : "{ default }",
            "from",
            `"${path}";`,
        ]
            .filter(Boolean)
            .join(" ")
    },
    defaultImport({ alias, path, type }: {
        alias: string
        path: string
        type?: boolean
    }): string {
        if (!path || !alias) return ""
        return [
            "import",
            type ? "type" : "",
            alias,
            "from",
            `"${path}";`,
        ]
            .filter(Boolean)
            .join(" ")
    },
    export(source: Source): string {
        if (!source.export) {
            return Barrel.wildcardExport({
                alias: source.alias,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        if (source.export.type === "Default") {
            return Barrel.defaultExport({
                alias: source.alias,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        if (source.export.type === "Named") {
            return Barrel.namedExport({
                alias: source.alias,
                name: source.export.identifier,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        if (source.export.type === "Wildcard") {
            return Barrel.wildcardExport({
                alias: source.alias,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        return ""
    },
    flat(sources: readonly Source[]): string {
        const types = sources
            .filter(source => Source.isTypeImport(source))
            .map(source => Barrel.export(source))
            .filter(Boolean)
            .sort((a, b) => a.localeCompare(b, "en-US"))

        const values = sources
            .filter(source => !Source.isTypeImport(source))
            .map(source => Barrel.export(source))
            .filter(Boolean)
            .sort((a, b) => a.localeCompare(b, "en-US"))

        return [...types, ...values].join("\n")
    },
    import(source: Source): string {
        if (!source.export) {
            return Barrel.wildcardImport({
                alias: source.alias,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        if (source.export.type === "Default") {
            return Barrel.defaultImport({
                alias: source.alias,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        if (source.export.type === "Named") {
            return Barrel.namedImport({
                alias: source.alias,
                name: source.export.identifier,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        if (source.export.type === "Wildcard") {
            return Barrel.wildcardImport({
                alias: source.alias,
                path: Source.importPath(source),
                type: Source.isTypeImport(source),
            })
        }

        return ""
    },
    isWatching(output: BarrelsOutput, path: string): boolean {
        path = Path.resolve(path)

        for (const matches of output.watchedPatterns.values()) {
            if (matches(path)) return true
        }

        return false
    },
    namedExport({ alias, name, path, type }: {
        alias?: string
        name: string
        path: string
        type?: boolean
    }): string {
        if (!name || !path) return ""
        return [
            "export",
            type ? "type" : "",
            "{",
            name,
            alias ? `as ${alias}` : "",
            "}",
            "from",
            `"${path}";`,
        ]
            .filter(Boolean)
            .join(" ")
    },
    namedImport({ alias, name, path, type }: {
        alias?: string
        name: string
        path: string
        type?: boolean
    }): string {
        if (!name || !path) return ""
        return [
            "import",
            type ? "type" : "",
            "{",
            name,
            alias ? `as ${alias}` : "",
            "}",
            "from",
            `"${path}";`,
        ]
            .filter(Boolean)
            .join(" ")
    },
    namespace(sources: readonly Source[], options: {
        indentation?: number
        name: string
        types?: "merge" | "nested" | "flat"
    }): string {
        const types = options.types || "flat"
        const candidates = sources.filter(source => Source.importName(source))
        const indentation = " ".repeat(options.indentation ?? 4)

        const typeImports = candidates
            .filter(source => Source.isTypeImport(source))
            .filter((source) => {
                if (types === "flat") return false
                if (types === "nested") return true
                return options.name === Source.importName(source)
            })
            .map(source => Source.setImportAlias(source, `_${Source.importName(source)}`))
            .map(source => Barrel.import(source))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const valueImports = candidates
            .filter(source => !Source.isTypeImport(source))
            .map(source => Source.setImportAlias(source, `_${Source.importName(source)}`))
            .map(source => Barrel.import(source))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const typeReExports = candidates
            .filter(source => Source.isTypeImport(source))
            .filter((source) => {
                if (types === "flat") return true
                if (types === "nested") return false
                return options.name !== Source.importName(source)
            })
            .map(source => Barrel.export(source))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const typeExports = candidates
            .filter(source => Source.isTypeImport(source))
            .filter((source) => {
                if (types === "nested") return false
                return options.name === Source.importName(source)
            })
            .map(source => Barrel.constExport({
                name: Source.importName(source),
                type: true,
                value: `_${Source.importName(source)}`,
            }))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const header = [
            [typeImports, valueImports].filter(Boolean).join("\n"),
            [typeReExports, typeExports].filter(Boolean).join("\n"),
        ].filter(Boolean).join("\n\n")

        const typeExportsNested = candidates
            .filter(source => Source.isTypeImport(source))
            .filter((source) => {
                if (types === "flat") return false
                if (types === "nested") return true
                return options.name !== Source.importName(source)
            })
            .map(source => Barrel.constExport({
                name: Source.importName(source),
                type: true,
                value: `_${Source.importName(source)}`,
            }))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .map(line => `${indentation}${line}`)
            .join("\n")

        const valueExports = candidates
            .filter(source => !Source.isTypeImport(source))
            .map(source => Barrel.constExport({
                name: Source.importName(source),
                value: `_${Source.importName(source)}`,
            }))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .map(line => `${indentation}${line}`)
            .join("\n")

        const body = [typeExportsNested, valueExports]
            .filter(Boolean)
            .join("\n")

        return `${header}\n\nexport namespace ${options.name} {\n${body}\n}`
    },
    property({ key, value }: {
        key: string
        value: string
    }): string {
        if (!key || !value) return ""
        if (key === value) return `${key},`
        return `${key}: ${value},`
    },
    record(sources: readonly Source[], options: {
        indentation?: number
        name: string
    }): string {
        const candidates = sources.filter(source => Source.importName(source))
        const indentation = " ".repeat(options.indentation ?? 4)

        const typeImports = candidates
            .filter(source => Source.isTypeImport(source))
            .filter(source => options.name === Source.importName(source))
            .map(source => Source.setImportAlias(source, `_${Source.importName(source)}`))
            .map(source => Barrel.import(source))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const valueImports = candidates
            .filter(source => !Source.isTypeImport(source))
            .map(source => Barrel.import(source))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const typeReExports = candidates
            .filter(source => Source.isTypeImport(source))
            .filter(source => options.name !== Source.importName(source))
            .map(source => Barrel.export(source))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const typeExports = candidates
            .filter(source => Source.isTypeImport(source))
            .filter(source => options.name === Source.importName(source))
            .map(source => Barrel.constExport({
                name: Source.importName(source),
                type: true,
                value: `_${Source.importName(source)}`,
            }))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .join("\n")

        const header = [
            [typeImports, valueImports].filter(Boolean).join("\n"),
            [typeReExports, typeExports].filter(Boolean).join("\n"),
        ].filter(Boolean).join("\n\n")

        const props = candidates
            .filter(source => !Source.isTypeImport(source))
            .map(source => Barrel.property({
                key: Source.importName(source),
                value: Source.importName(source),
            }))
            .sort((a, b) => a.localeCompare(b, "en-US"))
            .map(line => `${indentation}${line}`)
            .join("\n")

        return `${header}\n\nexport const ${options.name} = {\n${props}\n}`
    },
    watch(patterns: string | string[], basePath?: string): void {
        if (!Array.isArray(patterns)) return Barrel.watch([patterns], basePath)

        const output = getBarrelsOutput()
        if (!output) return

        patterns
            .filter(Boolean)
            .map(pattern => Path.resolve(basePath ?? process.cwd(), pattern))
            .forEach((pattern) => {
                if (output.watchedPatterns.has(pattern)) return
                output.watchedPatterns.set(pattern, createFilter(pattern))
            })
    },
    wildcardExport({ alias, path, type }: {
        alias?: string
        path: string
        type?: boolean
    }): string {
        if (!path) return ""
        return [
            "export",
            type ? "type" : "",
            "*",
            alias ? `as ${alias}` : "",
            "from",
            `"${path}";`,
        ]
            .filter(Boolean)
            .join(" ")
    },
    wildcardImport({ alias, path, type }: {
        alias: string
        path: string
        type?: boolean
    }): string {
        if (!path || !alias) return ""
        return [
            "import",
            type ? "type" : "",
            "*",
            alias ? `as ${alias}` : "",
            "from",
            `"${path}";`,
        ]
            .filter(Boolean)
            .join(" ")
    },
    async write(path: string, barrel: string): Promise<void> {
        // Add a trailing newline.
        barrel = `${barrel.trim()}\n`

        const targetDir = Path.dirname(path)

        // Check whether anything changed before writing, mostly there to prevent infinite loops in the file watcher.
        if (barrelsCache.get(path) === barrel) return
        barrelsCache.set(path, barrel)

        // Report to BarrelsRunner for pretty stdout outputs.
        const output = getBarrelsOutput()
        if (output) output.barrels.push({ barrel, path })

        mkdir(targetDir, { recursive: true })
        writeFile(path, barrel)
    },
}
