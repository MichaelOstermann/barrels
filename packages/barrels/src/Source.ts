import type { Node } from "oxc-parser"
import type { GlobOptions } from "tinyglobby"
import type { FilterPattern } from "unplugin-utils"
import type { SourceExport } from "./SourceModule"
import Path from "node:path"
import { glob } from "tinyglobby"
import { Barrel } from "./Barrel"
import { SourceModule } from "./SourceModule"

export interface Source {
    readonly alias: string
    readonly dirPath: string
    readonly export?: SourceExport
    readonly extName: string
    readonly isType: boolean
    readonly module: SourceModule
}

// eslint-disable-next-line @typescript-eslint/no-redeclare
export const Source = {
    async default(path: string, options?: {
        alias?: string
        type?: boolean
    }): Promise<Source | undefined> {
        const module = await SourceModule.resolve(path)
        if (!module) return

        return {
            alias: options?.alias ?? "",
            dirPath: module.dirPath,
            extName: module.extName,
            isType: false,
            module,
            export: {
                identifier: "",
                isType: options?.type ?? false,
                type: "Default",
            },
        }
    },
    async exports(source: Source | readonly Source[]): Promise<readonly Source[]> {
        if (isArray(source)) {
            const exports = await Promise.all(source.map(source => Source.exports(source)))
            return exports.flat(1)
        }

        if (source.alias) return [source]

        if (!source.export) {
            const exports = await SourceModule.exports(source.module, true)
            return exports.map(ex => ({ ...source, export: ex }))
        }

        if (source.export.type === "Wildcard" && source.export.from && !source.alias) {
            const exports = await SourceModule.exports(source.export.from, true)
            return exports.map(ex => ({ ...source, export: ex }))
        }

        return [source]
    },
    async file(path: string, options?: {
        alias?: string
        type?: boolean
    }): Promise<Source | undefined> {
        const module = await SourceModule.resolve(path)
            ?? await SourceModule.resolve(Path.resolve(path))

        if (!module) return

        return {
            alias: options?.alias ?? "",
            dirPath: module.dirPath,
            extName: module.extName,
            isType: options?.type ?? false,
            module,
        }
    },
    async files(
        pattern: string | string[],
        options?: Omit<GlobOptions, "absolute" | "patterns" | "onlyFiles" | "onlyDirectories">,
    ): Promise<Source[]> {
        Barrel.watch(pattern, options?.cwd)
        const files = await glob(pattern, {
            ...options,
            absolute: true,
            onlyFiles: true,
        })
        const sources = await Promise.all(files.map(path => Source.file(path)))
        return sources.filter(source => !!source)
    },
    importFrom<T extends Source>(source: T, filePath: string): T {
        if (source.module.isExternal) return source
        filePath = Path.resolve(filePath)
        const fromDir = Path.dirname(Path.resolve(filePath))
        const toDir = source.module.dirPath
        const relative = Path.relative(fromDir, toDir)
        const dirPath = relative.startsWith(".") ? relative : `./${relative}`
        return { ...source, dirPath }
    },
    importName(source: Source): string {
        return source.alias || source.export?.identifier || ""
    },
    importPath(source: Source): string {
        if (source.module.isExternal)
            return source.module.fileName

        if (Path.isAbsolute(source.dirPath))
            source = Source.importFrom(source, "./index.ts")

        const extName = source.extName ? `.${source.extName}` : ""

        return [
            ...source.dirPath.split(Path.sep),
            `${source.module.fileName}${extName}`,
        ]
            .filter(Boolean)
            .join(Path.sep)
    },
    isTypeExport(source: Source): boolean {
        return source.export?.isType ?? false
    },
    isTypeImport(source: Source): boolean {
        return source.export?.isType ?? source.isType
    },
    matches(source: Source, pattern: Exclude<FilterPattern, null>): boolean {
        return SourceModule.matches(source.module, pattern)
    },
    async named(name: string, path: string, options?: {
        alias?: string
        type?: boolean
    }): Promise<Source | undefined> {
        const module = await SourceModule.resolve(path)
        if (!module) return

        return {
            alias: options?.alias ?? "",
            dirPath: module.dirPath,
            extName: module.extName,
            isType: false,
            module,
            export: {
                identifier: name,
                isType: options?.type ?? false,
                type: "Named",
            },
        }
    },
    remapExtension<T extends Source>(source: T, mapping: Record<string, string>): T {
        if (source.module.isExternal) return source
        if (!source.extName) return source
        if (!(source.extName in mapping)) return source
        const extName = mapping[source.extName] || ""
        return { ...source, extName }
    },
    removeExtension<T extends Source>(source: T): T {
        if (source.module.isExternal) return source
        return { ...source, extName: "" }
    },
    setImportAlias<T extends Source>(source: T, alias: string): T {
        return { ...source, alias }
    },
    toTypeImport<T extends Source>(source: T): T {
        return { ...source, isType: true }
    },
    walkAst(source: Source, options: {
        enter: (node: Node) => void | Promise<void>
        leave: (node: Node) => void | Promise<void>
    }): Promise<void> {
        return SourceModule.walkAst(source.module, options)
    },
    async wildcard(path: string, options?: {
        alias?: string
        type?: boolean
    }): Promise<Source | undefined> {
        const module = await SourceModule.resolve(path)
        if (!module) return

        return {
            alias: options?.alias ?? "",
            dirPath: module.dirPath,
            extName: module.extName,
            isType: false,
            module,
            export: {
                identifier: "",
                isType: options?.type ?? false,
                type: "Wildcard",
            },
        }
    },
}

function isArray(value: unknown): value is readonly unknown[] {
    return Array.isArray(value)
}
