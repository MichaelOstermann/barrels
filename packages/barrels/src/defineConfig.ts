import type { BarrelConfig, BarrelsOutput, CustomBarrelConfig } from "./types"
import { Barrel } from "./Barrel"
import { runConfig } from "./internals/context"
import { resolveBarrel } from "./internals/resolveBarrel"
import { resolveDirectories } from "./internals/resolveDirectories"
import { resolveOutput } from "./internals/resolveOutput"
import { resolveSources } from "./internals/resolveSources"

type Config =
    | BarrelConfig
    | CustomBarrelConfig
    | (() => Promise<BarrelsOutput>)

export function defineConfig(
    configs: Config[],
): () => Promise<BarrelsOutput> {
    return () => runConfig(async () => {
        for (const config of configs) {
            try {
                if (typeof config === "function") {
                    await config()
                }
                else {
                    const directories = await resolveDirectories(config)
                    await Promise.all(directories.map(async (directory) => {
                        const output = resolveOutput(directory, config)
                        const sources = await resolveSources(directory, output, config)
                        const barrel = await resolveBarrel(output, sources, config)
                        await Barrel.write(output, barrel)
                    }))
                }
            }
            catch (error) {
                console.error(error)
            }
        }
    })
}
