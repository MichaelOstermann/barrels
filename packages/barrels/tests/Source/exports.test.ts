import { describe, expect, it } from "vitest"
import { Source } from "../../src"

describe("Source.exports", () => {
    it("should retrieve exports from local module", async () => {
        const file = await Source.file("./packages/barrels/tests/__fixtures__/exports.ts")
        const exports = await Source.exports(file!)
        expect(exports).toMatchSnapshot()
    })

    it("should retrieve exports from external module", async () => {
        const file = await Source.file("tsconfck")
        const exports = await Source.exports(file!)
        expect(exports).toMatchSnapshot()
    })
})
