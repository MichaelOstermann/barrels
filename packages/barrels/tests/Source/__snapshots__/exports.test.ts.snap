// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Source.exports > should retrieve exports from external module 1`] = `
[
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./find.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/find.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "find",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./find-all.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/find-all.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "findAll",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./to-json.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/to-json.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "toJson",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./parse.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/parse.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "parse",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./parse.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/parse.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "TSConfckParseError",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./find-native.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/find-native.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "findNative",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./parse-native.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/parse-native.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "parseNative",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./parse-native.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/parse-native.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "TSConfckParseNativeError",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
    "export": {
      "from": {
        "allowImportingTsExtensions": undefined,
        "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
        "extName": "js",
        "fileName": "./cache.js",
        "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/cache.js",
        "isExternal": true,
        "moduleResolution": undefined,
        "tsconfigPath": "",
      },
      "identifier": "TSConfckCache",
      "isType": false,
      "type": "Named",
    },
    "extName": "js",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/node_modules/tsconfck/src",
      "extName": "js",
      "fileName": "tsconfck",
      "filePath": "/Users/michael/Development/barrels/node_modules/tsconfck/src/index.js",
      "isExternal": true,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/tsconfig.json",
    },
  },
]
`;

exports[`Source.exports > should retrieve exports from local module 1`] = `
[
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
    "export": {
      "identifier": "",
      "isType": false,
      "type": "Default",
    },
    "extName": "ts",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
      "extName": "ts",
      "fileName": "exports",
      "filePath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/exports.ts",
      "isExternal": false,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
    "export": {
      "identifier": "A",
      "isType": false,
      "type": "Named",
    },
    "extName": "ts",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
      "extName": "ts",
      "fileName": "exports",
      "filePath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/exports.ts",
      "isExternal": false,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
    "export": {
      "identifier": "B",
      "isType": false,
      "type": "Named",
    },
    "extName": "ts",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
      "extName": "ts",
      "fileName": "exports",
      "filePath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/exports.ts",
      "isExternal": false,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
    "export": {
      "identifier": "C",
      "isType": false,
      "type": "Named",
    },
    "extName": "ts",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
      "extName": "ts",
      "fileName": "exports",
      "filePath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/exports.ts",
      "isExternal": false,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
    "export": {
      "identifier": "D",
      "isType": false,
      "type": "Named",
    },
    "extName": "ts",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
      "extName": "ts",
      "fileName": "exports",
      "filePath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/exports.ts",
      "isExternal": false,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
    "export": {
      "identifier": "E",
      "isType": true,
      "type": "Named",
    },
    "extName": "ts",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
      "extName": "ts",
      "fileName": "exports",
      "filePath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/exports.ts",
      "isExternal": false,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/tsconfig.json",
    },
  },
  {
    "alias": "",
    "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
    "export": {
      "identifier": "F",
      "isType": true,
      "type": "Named",
    },
    "extName": "ts",
    "isType": false,
    "module": {
      "allowImportingTsExtensions": undefined,
      "dirPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__",
      "extName": "ts",
      "fileName": "exports",
      "filePath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/exports.ts",
      "isExternal": false,
      "moduleResolution": "bundler",
      "tsconfigPath": "/Users/michael/Development/barrels/packages/barrels/tests/__fixtures__/tsconfig.json",
    },
  },
]
`;
